### PR TITLE
Use Sphinx autodoc to create the API reference instead of Epydoc

### PR DESCRIPTION
**Pull request by cjmayo**
Epydoc is Python 2 only and unmaintained.

---

I've had a go at producing the API reference with Sphinx. This is the output of sphinx-apidoc tweaked a little bit. Plus a few other commits that would probably be useful anyway.

To make for an easier review I've published it at: https://cjmayo.github.io/lxml/api/.

I've tried to create a straightforward replacement. I have updated the PDF generation, but not tested because I don't have all the LaTeX files installed.

### CODE REVIEW

File: src/lxml/pyclasslookup.py [hunk 1]
```diff
@@ -1,3 +1,3 @@
-# dummy module for backwards compatibility
+"""Dummy module for backwards compatibility."""
```
Comment (scoder) [inline src/lxml/pyclasslookup.py:1]: That suggests that it's not worth documenting.

**Comment by bkline** [comment]
It certainly looks prettier, but if this is a replacement for https://lxml.de/api/ I'm not finding everything. It's possible I'm not looking in the right places, but where for example is documentation for the members of `_Element`?

File: doc/api/lxml.tests.rst [hunk 2]
```diff
@@ -0,0 +1,282 @@
+lxml.tests
```
Comment (scoder) [inline doc/api/lxml.tests.rst:1]: I think tests were previously only included by accident. I think it's useless to add them here.

File: doc/api/lxml.html.rst [hunk 3]
```diff
@@ -0,0 +1,81 @@
+lxml.html
```
Comment (scoder) [inline doc/api/lxml.html.rst:1]: Isn't there a way to generate these index files automatically? They seem trivial, and that's unlikely to change. But when new modules or packages are added, it's annoying to have to remember and add them here.

**Comment by scoder** [comment]
> where for example is documentation for the members of `_Element` ?

Yeah, "private" classes in `lxml` are not all that private, really. They are not intended to be instantiated by users, but they are part of the API and need to be included in the API documentation.

File: src/lxml/pyclasslookup.py (hunk 1 above)
Comment (cjmayo) [reply src/lxml/pyclasslookup.py]: OK. Reverted and dropped from the documentation.

File: doc/api/lxml.tests.rst (hunk 2 above)
Comment (cjmayo) [reply doc/api/lxml.tests.rst]: Removed tests.

File: doc/api/lxml.html.rst (hunk 3 above)
Comment (cjmayo) [reply doc/api/lxml.html.rst:1]: There is `sphinx-autogen` [1] which can generate module rst files based on one simple top-level rst. You have to have summaries, which maybe is OK, but I can't see how to get the summaries to include private classes like lxml.etree._Element.

[1] https://www.sphinx-doc.org/en/master/man/sphinx-autogen.html

**Comment by cjmayo** [comment]
> I think there isn't anything in the travis setup currently that builds the docs, though. Probably worth adding a step for that as well – or a dedicated build target?

Added a `make html` to Travis.

I've just added commits here, I don't know what your preference is for tidying up with rebasing and squashing.

**Comment by cjmayo** [comment]
> > where for example is documentation for the members of `_Element` ?
> 
> Yeah, "private" classes in `lxml` are not all that private, really. They are not intended to be instantiated by users, but they are part of the API and need to be included in the API documentation.

Needs `:ignore-module-all:` as well I think because `_Element` is not in `lxml.etree.__all__`. Added and moved down to extensions section.

File: doc/api/lxml.html.rst (hunk 3 above)
Comment (scoder) [reply doc/api/lxml.html.rst:1]: Does it not take the default "autodoc" options into account?

Comment (cjmayo) [reply doc/api/lxml.html.rst:1]: In the autodoc sections yes, but not in the autosummary tables



**Classes**


&nbsp; | &nbsp;
-- | --
AncestorsIterator(self, node[, tag]) | Iterates over the ancestors of an element (from parent to parent).
AttributeBasedElementClassLookup(self, …) | Checks an attribute of an Element and looks up the value in a class dictionary.
C14NWriterTarget | Canonicalization writer target for the XMLParser.
CDATA(data) | CDATA factory.
...

Comment (cjmayo) [reply doc/api/lxml.html.rst:1]: Maybe it is possible to use `sphinx-apidoc` every time and forget about the summaries, it does take quite a few options, including `--private`.

Comment (cjmayo) [reply doc/api/lxml.html.rst:1]: > Maybe it is possible to use `sphinx-apidoc` every time and forget about the summaries, it does take quite a few options, including `--private`.

Created #309 using this approach.

**Comment by scoder** [comment]
Closing, superseded by #309.

**Pull request closed without merging**